nvflare/app_common/app_defined/shareable_generator.py

```diff
@@ -0,0 +1,94 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+from typing import Any
+
+from nvflare.apis.dxo import DXO, DataKind, from_shareable
+from nvflare.apis.fl_context import FLContext
+from nvflare.apis.shareable import Shareable
+from nvflare.app_common.abstract.learnable import Learnable
+from nvflare.app_common.abstract.model import ModelLearnable, ModelLearnableKey, make_model_learnable
+from nvflare.app_common.abstract.shareable_generator import ShareableGenerator
+from nvflare.app_common.app_constant import AppConstants
+
+from .component_base import ComponentBase
+
+
+class AppDefinedShareableGenerator(ShareableGenerator, ComponentBase, ABC):
+    def __init__(self):
+        ShareableGenerator.__init__(self)
+        ComponentBase.__init__(self)
+        self.current_round = None
+
+    @abstractmethod
+    def model_to_trainable(self, model_obj: Any) -> (Any, dict):
+        """Convert the model weights and meta to a format that can be sent to clients to do training
+
+        Args:
+            model_obj: model object
+
+        Returns: a tuple of (weights, meta)
+
+        The returned weights and meta will be for training and serializable
+        """
+        pass
+
+    @abstractmethod
+    def apply_weights_to_model(self, model_obj: Any, weights: Any, meta: dict) -> Any:
+        """Apply trained weights and meta to the base model
+
+        Args:
+            model_obj: base model object that weights will be applied to
+            weights: trained weights
+            meta: trained meta
+
+        Returns: the updated model object
+
+        """
+        pass
```

Comment:
We can discuss this with AV / users about what interface they want?

I personally think apply_weights_to_model sounds like an aggregator job?
Generator should be "model_to_trainable" and "trainable_to_model"?
What is even better if user don't need to write this class?

Reply:
I debated with myself on this. I think "apply_weights_to_model" is a better name here: the model could have a lot more info than just the weights. What this method does is to apply trained result to the model (and leave other things of the model intact).

Maybe we should change the name to "apply_training_result_to_model", instead of calling the training_result "weights".